TestCentricEngine.md

```diff
@@ -0,0 +1,82 @@
+ * **nunit-agent.exe** and **nunit-agent-x86.exe** are console apps that serve as cross-processs agents when invoked by the engine. They reference the engine code and use configuration options to enable only those parts of the engine assembly that are intended for use by an agent.
+
+It may seem odd that the NUnit engine assembly is used for two purposes: to function as an engine and to provide services as part of an agent process. This was initially done for expediency, with the plan of splitting the code into two assemblies, one related to engine functions _per se_ and the other containing the code needed by an agent. Unfortunately, this is a rather large amount of technical debt that was never paid off. 
+
+## TestCentric GUI Requirements
+
+Currently, the TestCentric GUI is using a modified build of the NUnit engine. So far, the modifications are minor. In this section, I want to examine some key features we will need in the near future, which are not currently provided by the NUnit engine.
+
+1. The Gui should be able to run tests on other platforms than .NET Framework, such as .NET Core. It should, in fact, be able to load and run multiple assemblies that use different platforms. Currently, that's not possible because it requires using two different builds of the engine in the same process.
+
+2. The GUI chocolatey package should reference the engine package, rather than including all the engine assemblies in it's own package. This is not currently possible because of how the engine itself is packaged.
+
+3. The engine should support extensions to extensions. An issue exists for this in the NUnit project.
+
+4. The engine should return proper summary tests for each supported project that is run. An issue exists for this in the NUnit project. Currently, there is adhoc code in the GUI to add this information to the visual display but it is, of course, not present in the result XML.
+
+5. In order to communicate between the engine and agents running under multiple platforms and runtimes, new communications channels are needed. Currently, the engine only communicates across processes and AppDomains using .NET Remoting. There is an existing issue for this in the NUnit project.
+
+6. The engine should support extensions to the GUI itself as well as to other runners. This was part of the original architectural design and requires the ExtensionService to examine multiple roots (assemblies) rather than just the nunit.engine and nunit.engine.api assemblies. It calls for an API whereby the user can specify the roots to be examined and, in turn, will necessitate a more dynamic approach to loading and unloading extensions. Currently, extension points and extensions are detected once, upon initialization of the service, and are not allowed to change in the course of a run. There is no current issue in the NUnit project about this.
+
```

Comment:
Interested in this one. There's been a couple of issues which relate to NUnit Console extensions, and some 'debt' which would be remedied by such things too (e.g. the console's knowledge of the `--teamcity` flag)

Not something I've previously put a lot of thought in to however.

Reply:
Yeah... the original design (c. 2007) was based on use of `Mono.Addins`, which supported multiply rooted trees of extensions for exactly that reason. When that package became unusable (they abandoned support for .NET 2.0) I developed a minimal extension package, the one we now have. Allowing for multiple roots __statically__ is actually not that big a change to make. The dynamic part is a bit harder - i.e. allowing the engine to initialize itself, followed by the runner adding it's own initialization on top of that.